Source/Core/AudioCommon/CubebUtils.cpp

```diff
@@ -82,3 +91,146 @@ std::shared_ptr<cubeb> CubebUtils::GetContext()
   weak = shared = {ctx, DestroyContext};
   return shared;
 }
+
+std::vector<std::pair<std::string, std::string>> ListInputDevices()
+{
+  std::vector<std::pair<std::string, std::string>> devices;
+
+  cubeb_device_collection collection;
+  auto cubeb_ctx = GetContext();
+  int r = cubeb_enumerate_devices(cubeb_ctx.get(), CUBEB_DEVICE_TYPE_INPUT, &collection);
+
+  if (r != CUBEB_OK)
+  {
+    ERROR_LOG_FMT(AUDIO, "Error listing cubeb input devices");
+    return devices;
+  }
+
+  INFO_LOG_FMT(AUDIO, "Listing cubeb input devices:");
+  for (uint32_t i = 0; i < collection.count; i++)
+  {
+    auto& info = collection.device[i];
+    auto& device_state = info.state;
+    const char* state_name = [device_state] {
+      switch (device_state)
+      {
+      case CUBEB_DEVICE_STATE_DISABLED:
+        return "disabled";
+      case CUBEB_DEVICE_STATE_UNPLUGGED:
+        return "unplugged";
+      case CUBEB_DEVICE_STATE_ENABLED:
+        return "enabled";
+      default:
+        return "unknown?";
+      }
+    }();
+
+    INFO_LOG_FMT(AUDIO,
+                 "[{}] Device ID: {}\n"
+                 "\tName: {}\n"
+                 "\tGroup ID: {}\n"
+                 "\tVendor: {}\n"
+                 "\tState: {}",
+                 i, info.device_id, info.friendly_name, info.group_id,
```

Comment:
This throws an exception on my system from `info.group_id` being `nullptr`. Maybe all these strings need to be checked like you did with `vendor_name`.

Reply:
I added more sanity checks to prevent it from happening. Sadly, I wasn't expecting a nullptr here as it wasn't mentioned in the struct description in cubeb.h :v/.